server.go: use worker goroutines for fewer stack allocations

adtac opened the pull request:
Currently (go1.13.4), the default stack size for newly spawned
goroutines is 2048 bytes. This is insufficient when processing gRPC
requests as the we often require more than 4 KiB stacks. This causes the
Go runtime to call runtime.morestack at least twice per RPC, which
causes performance to suffer needlessly as stack reallocations require
all sorts of internal work such as changing pointers to point to new
addresses.

Since this stack growth is guaranteed to happen at least twice per RPC,
reusing goroutines gives us two wins:

  1. The stack is already grown to 8 KiB after the first RPC, so
     subsequent RPCs do not call runtime.morestack.
  2. We eliminate the need to spawn a new goroutine for each request
     (even though they're relatively inexpensive).

Performance improves across the board. The improvement is especially
visible in small, unary requests as the overhead of stack reallocation
is higher, percentage-wise. QPS is up anywhere between 3% and 5%
depending on the number of concurrent RPC requests in flight. Latency is
down ~3%. There is even a 1% decrease in memory footprint in some cases,
though that is an unintended, but happy coincidence.

```
unary-networkMode_none-bufConn_false-keepalive_false-benchTime_1m0s-trace_false-latency_0s-kbps_0-MTU_0-maxConcurrentCalls_8-reqSize_1B-respSize_1B-compressor_off-channelz_false-preloader_false
               Title       Before        After Percentage
            TotalOps      2613512      2701705     3.37%
             SendOps            0            0      NaN%
             RecvOps            0            0      NaN%
            Bytes/op      8657.00      8654.17    -0.03%
           Allocs/op       173.37       173.28     0.00%
             ReqT/op    348468.27    360227.33     3.37%
            RespT/op    348468.27    360227.33     3.37%
            50th-Lat    174.601µs    167.378µs    -4.14%
            90th-Lat    233.132µs    229.087µs    -1.74%
            99th-Lat     438.98µs    441.857µs     0.66%
             Avg-Lat    183.263µs     177.26µs    -3.28%
```

adtac commented on server.go:
```diff
@@ -712,15 +712,53 @@ func (s *Server) newHTTP2Transport(c net.Conn, authInfo credentials.AuthInfo) tr
+	st.HandleStreams(func(stream *transport.Stream) {
+		wg.Add(1)
+		select {
+		case streamChannels[stream.ID() & workerMask] <- stream:
```
If you have a cheaper source of pseudo-randomness, please let me know :)

adtac commented on server.go:
```diff
@@ -712,15 +712,66 @@ func (s *Server) newHTTP2Transport(c net.Conn, authInfo credentials.AuthInfo) tr
 	return st
 }
 
+func floorCPUCount() uint32 {
```
this is code repetition from the profiling core PR. I don't know where something like this should live so that it can be reused by both `google.golang.org/grpc` and `google.golang.org/grpc/internal/profiling` - any suggestions?

adtac commented:
![qps](https://user-images.githubusercontent.com/7521600/69388752-cfcdb900-0c7e-11ea-831e-ec083af9b1ad.png)
![50th](https://user-images.githubusercontent.com/7521600/69388758-d0fee600-0c7e-11ea-88e4-d0e890849ef1.png)
![bytes](https://user-images.githubusercontent.com/7521600/69388764-d3f9d680-0c7e-11ea-934c-75f08b459fdb.png)

dfawley commented:
Per offline discussions:
- Please add a ServerOption to configure this value, disabled by default
- Benchmark the difference between using `%` instead of `&`.
- Use `streamID >> 1` instead of `streamID` since the client always uses odd stream IDs.
- Run benchmarks to compare this approach with the more obvious one that uses a single channel with multiple listeners and report the results.

adtac commented:
```diff
commit e761da0c0417e4d0855226bc9f7169eccb85658f
Author: Adhityaa Chandrasekar <adtac@google.com>
Date:   Wed Dec 18 15:05:29 2019 -0800

    V2.1: single channel

diff --git a/server.go b/server.go
index dec8ab2f..2fe14ed5 100644
--- a/server.go
+++ b/server.go
@@ -759,20 +759,19 @@ func (s *Server) serveStreams(st transport.ServerTransport) {
 	defer st.Close()
 	var wg sync.WaitGroup
 
-	var streamChannels []chan *transport.Stream
-	if s.numStreamWorkers > 0 {
-		streamChannels = make([]chan *transport.Stream, numStreamWorkers)
-		for i := range streamChannels {
-			streamChannels[i] = make(chan *transport.Stream)
-			go s.streamWorker(st, &wg, streamChannels[i])
+	var streamChannel chan *transport.Stream
+	if s.opts.numStreamWorkers > 0 {
+		streamChannel = make(chan *transport.Stream)
+		for i := uint32(0); i < s.opts.numStreamWorkers; i++ {
+			go s.streamWorker(st, &wg, streamChannel)
 		}
 	}
 
 	st.HandleStreams(func(stream *transport.Stream) {
 		wg.Add(1)
-		if s.numStreamWorkers > 0 {
+		if s.opts.numStreamWorkers > 0 {
 			select {
-			case streamChannels[(stream.ID() >> 1) % s.numStreamWorkers] <- stream:
+			case streamChannel <- stream:
 			default:
 				// If all stream workers are busy, fallback to default code path.
 				go func() {
@@ -795,10 +794,8 @@ func (s *Server) serveStreams(st transport.ServerTransport) {
 	})
 	wg.Wait()
 
-	if s.numStreamWorkers > 0 {
-		for _, ch := range streamChannels {
-			close(ch)
-		}
+	if s.opts.numStreamWorkers > 0 {
+		close(streamChannel)
 	}
 }
 ```

```
unary-networkMode_none-bufConn_false-keepalive_false-benchTime_1m0s-trace_false-latency_0s-kbps_0-MTU_0-maxConcurrentCalls_1-reqSize_1B-respSize_1B-compressor_off-channelz_false-preloader_false
               Title       Before        After Percentage
            TotalOps       393011       389889    -0.79%
             SendOps            0            0      NaN%
             RecvOps            0            0      NaN%
            Bytes/op      9527.99      9527.14     0.00%
           Allocs/op       224.72       224.63     0.00%
             ReqT/op     52401.47     51985.20    -0.79%
            RespT/op     52401.47     51985.20    -0.79%
            50th-Lat    147.933µs     148.88µs     0.64%
            90th-Lat    183.161µs     183.69µs     0.29%
            99th-Lat    297.636µs    298.317µs     0.23%
             Avg-Lat     152.14µs    153.361µs     0.80%
```

~0.79% hit using just one channel - I can't recollect if it was the same the last time I tried this, but the difference wasn't similarly small, but significant. Ran the test for 60s.

adtac commented:
>Please add a ServerOption to configure this value, disabled by default

Done.

>Benchmark the difference between using % instead of &.

Turns out to be insignificant. Also, since we're now allowing the user to set the number of stream workers, enforcing it to be an exponent of two is probably not a good idea. Switched to modulo.

>Use streamID >> 1 instead of streamID since the client always uses odd stream IDs.

I just realised that there's a slight issue with using stream ID -- it's biased towards lower numbers. If a server receives one unary request from several clients (a common workload), they'll all go to the same channel because the first stream ID is always 0.

Switched to a round-robin method that's actually fairer and 1% faster (~compared to master, round robin is a 5.33% improvement while stream ID is a ~4.54% improvement).

>Run benchmarks to compare this approach with the more obvious one that uses a single channel with multiple listeners and report the results.

See previous comment.

Pushed V2.

dfawley commented on internal/transport/transport.go:
```diff
@@ -286,6 +286,11 @@ type Stream struct {
 }
 
+// ID returns the stream ID.
+func (s *Stream) ID() uint32 {
```
Delete

adtac replied:
Done.

dfawley commented on server.go:
```diff
@@ -712,15 +729,64 @@ func (s *Server) newHTTP2Transport(c net.Conn, authInfo credentials.AuthInfo) tr
+	}
+}
+
 func (s *Server) serveStreams(st transport.ServerTransport) {
```
The pool will be created per-transport (connection) when it is done here.

You could have a `sync.Once` in `Server` to initialize them instead, and reap them in `Server.Stop`, if you want it to be per-grpc-server.

adtac replied:
Good idea, switched to server worker goroutines. Done.

adtac replied:
Not entirely done... a couple of tests failing.

aaronbee commented on server.go:
```diff
@@ -388,6 +397,56 @@ func HeaderTableSize(s uint32) ServerOption {
+//
+// [1] https://github.com/golang/go/issues/18138
+func (s *Server) serverWorker() {
+	for completed := 0; completed < serverWorkerResetThreshold; completed++ {
```
A random chance to reset the goroutine would probably be better than a fixed number of requests handled. The problem with fixed number of requests is that it's probable all of your goroutines will reset very close to each other in time if the load is evenly distributed among them. If they reset all at the same time you would have worse worst-case latency than if they each reset at different times.

adtac replied:
good catch, thanks! I've now changed it so that the threshold is randomly between `serverWorkerResetThreshold` and `2*serverWorkerResetThreshold`.

dfawley commented on server.go:
```diff
@@ -388,6 +398,59 @@ func HeaderTableSize(s uint32) ServerOption {
+	})
+}
+
+// serverWorkerResetThreshold defines how often the stack must be reset. Every
+// N requests, by spawning a new goroutine in its place, a worker can reset its
+// stack so that large stacks don't live in memory forever. 2^16 should allow
+// each goroutine stack to live for at least a few seconds in a typical
+// workload (assuming a QPS of a few thousand requests/sec).
+const serverWorkerResetThreshold = 1 << 16
```
Brainstorming: should this be time-based instead of request-based?  Or some combination of both?

If a server goes idle, we may want to restart the threads.

adtac replied:
Possibly yes, but I expect the overhead introduced by that to be non-negligible. Also heavily depends on how Go's runtime manages shrinking stacks during GC pauses, I think.

dfawley replied:
I think we could do it with a `Timer` checked in the same `select` as the workload without being too expensive.  But this is fine for now - we can work on this more in the future.

CAFxX replied:
a late comment just to point out that the runtime shrinks the stacks of goroutines during GC (or at the next safepoint): https://github.com/golang/go/blob/9d812cfa5cbb1f573d61c452c864072270526753/src/runtime/mgcmark.go#L781-L783

dfawley commented:
(Still interested in seeing some new results after the latest changes, otherwise LGTM.)

adtac commented:
ok this is weird, I'm now getting a 15% perf improvement -- why's there a sudden jump at 16? I think it's because I was using 16 worker goroutines.

![50th](https://user-images.githubusercontent.com/7521600/71297103-f42cbc00-2336-11ea-8941-f6f596d88570.png)
![qps](https://user-images.githubusercontent.com/7521600/71297104-f42cbc00-2336-11ea-8884-4e085972dbe5.png)

CAFxX commented on server.go:
```diff
@@ -715,12 +789,27 @@ func (s *Server) newHTTP2Transport(c net.Conn, authInfo credentials.AuthInfo) tr
+			default:
+				// If all stream workers are busy, fallback to the default code path.
+				go func() {
+					s.handleStream(st, stream, s.traceInfo(st, stream))
+					wg.Done()
```
this should have been 

```golang
defer wg.Done()
s.handleStream(st, stream, s.traceInfo(st, stream))
```

as it is on line 809 and as it was before the change

CAFxX commented on server.go:
```diff
@@ -715,12 +789,27 @@ func (s *Server) newHTTP2Transport(c net.Conn, authInfo credentials.AuthInfo) tr
+		if s.opts.numServerWorkers > 0 {
+			data := &serverWorkerData{st: st, wg: &wg, stream: stream}
+			select {
+			case s.serverWorkerChannels[atomic.AddUint32(&roundRobinCounter, 1)%s.opts.numServerWorkers] <- data:
```
this is potentially pretty suboptimal in case some workers are busy on requests that take significant time

CAFxX commented on server.go:
```diff
@@ -715,12 +789,27 @@ func (s *Server) newHTTP2Transport(c net.Conn, authInfo credentials.AuthInfo) tr
+			select {
+			case s.serverWorkerChannels[atomic.AddUint32(&roundRobinCounter, 1)%s.opts.numServerWorkers] <- data:
+			default:
+				// If all stream workers are busy, fallback to the default code path.
```
I think we should add a worker in this case, not a one-off goroutine.

Extra workers created over numServerWorkers should linger for a short period of time waiting for new work, and then shut down if no new work arrives.
